pkg/private/pkg_files.bzl

```diff
@@ -0,0 +1,305 @@
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#    http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""Internal methods for processing pkg_file* instances.
+
+Concepts and terms:
+
+  DestFile: A provider holding the source path, attributes and other
+            information about a file that should appear in the package.
+            When attributes are empty in DestFile, we let the package
+            tool decided their values.
+
+  content map: The map of destination paths to DestFile instances. Note that
+               several distinct destinations make share the same source path.
+               Attempting to insert a duplicate entry in the content map is
+               an error, because it means you are collapsing files together.
+               We may want to relax this in the future if their DestFiles
+               are equal.
+
+  manifest: The file which represents the content map. This is generated
+            by rule implementations and passed to the build_*.py helpers.
+"""
+
+load("//:path.bzl", "compute_data_path", "dest_path")
+load(
+    "//:providers.bzl",
+    "PackageArtifactInfo",
+    "PackageDirsInfo",
+    "PackageFilegroupInfo",
+    "PackageFilesInfo",
+    "PackageSymlinkInfo",
+)
+
+_DestFile = provider(
+    doc = """Information about each destination in the final package.""",
+    fields = {
+        "src": "source file",
+        "mode": "mode, or empty",
+        "user": "user, or empty",
+        "group": "group, or empty",
+        "link_to": "path to link to. src must not be set",
+        "is_dir": "int. if 1, create empty dir at dest. If 2 src" +
+                  " is a the top of a tree artifact to place at dest",
+        "origin": "target which added this",
+    },
+)
+
+def _check_dest(content_map, dest, origin):
+    if dest in content_map:
+        fail("Duplicate output path: <%s>, declared in %s and %s" % (
+            dest,
+            origin,
+            content_map[dest].origin,
+        ))
+
+def _merge_attributes(info, mode, user, group):
+    if hasattr(info, "attributes"):
+        attrs = info.attributes
+        mode = attrs.get("mode") or mode
+        user = attrs.get("user") or user
+        group = attrs.get("group") or group
+    return (mode, user, group)
+
+def _process_pkg_dirs(content_map, pkg_dirs_info, origin, default_mode, default_user, default_group):
+    attrs = _merge_attributes(pkg_dirs_info, default_mode, default_user, default_group)
+    for dir in pkg_dirs_info.dirs:
+        dest = dir.strip('/')
```

Comment:
That should happen in a different way. I want the interchange format to always be rootless. That makes reasoning easier for build_*.py.  They can take a separate parameter to reroot the entire tree.

Reply:
I see.  The question, then, is where would this rerooting happen, and how it would be specified?  My original intent was that this was only allowed in grouping (`pkg_filegroup`) rules.

As currently written, this isn't possible.  `_process_pkg_filegroup` delegates to the individual provider processing functions, which currently unconditionally strip the `/` prefix.

I'm fine with this going in as-is for now, but we need to come to a consensus on how this is intended to be dealt with.